### PR TITLE
Chasm/hole turfs

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
You step on it and you fall through. It drops you on the same X/Y on the next Z level. If you're on the last Z level it just deletes you instead.

For use in lavaworld related things (surface Z corresponding with an underground Z). 

TODO: some way to build on them

Sprites are from WJ

lava

### CODE REVIEW

File: code/modules/mining/mine_turfs.dm [hunk 1]
```diff
@@ -640,6 +640,43 @@ var/global/list/rockTurfEdgeCache
...
+	name = "chasm"
+	desc = "Watch your step."
+	baseturf = /turf/simulated/chasm
+	smooth = SMOOTH_TRUE
+	icon = 'icons/turf/floors/Chasms.dmi'
+	icon_state = "smooth"
+
+
+/turf/simulated/chasm/Entered(atom/movable/AM)
+	if(istype(AM, /obj/singularity) || istype(AM, /obj/item/projectile))
```
Comment (Incoming5643) [inline code/modules/mining/mine_turfs.dm]: You should also whitelist things that in the middle of being thrown

Comment (KorPhaeron) [inline code/modules/mining/mine_turfs.dm]: Whats the var for that?

Comment (Incoming5643) [inline code/modules/mining/mine_turfs.dm]: throwing

Comment (KorPhaeron) [inline code/modules/mining/mine_turfs.dm]: Thanks

**Comment by Bawhoppen** [comment]
slowly but surely

**Comment by Incoming5643** [comment]
Also some mobs obviously can fly, but I don't think that's tracked by a var in code anywhere.

Oh also, this could del ghosts.

**Comment by Alek2ander** [comment]
https://www.youtube.com/watch?v=M2MZ-4RWJR0

**Comment by WJohn** [comment]
It's a little unintuitive to know when you're at the bottom most z level and when you'll get deleted, but I think that'd be more a mapping issue than anything. This could be resolved by simply not using these on the bottom-most z level.

Speaking of that, how would these work on something like the caves away mission? There's four "floors" but they're all on the same z level.

File: code/modules/mining/mine_turfs.dm [hunk 2]
```diff
@@ -640,6 +640,43 @@ var/global/list/rockTurfEdgeCache
...
+		return
+	drop(AM)
+
+
+/turf/simulated/chasm/proc/drop(atom/movable/AM)
+	visible_message("[AM] falls into [src]!")
+	if(z+1 > world.maxz)
+		qdel(AM)
+	else
+		AM.Move(locate(src.x, src.y, src.z+1))
```
Comment (RemieRichards) [inline code/modules/mining/mine_turfs.dm]: just make these vars
```drop_x```,```drop_y```,```drop_z```

File: code/modules/mining/mine_turfs.dm [hunk 3]
```diff
@@ -640,6 +640,43 @@ var/global/list/rockTurfEdgeCache
...
+
+/turf/simulated/chasm/Entered(atom/movable/AM)
+	if(istype(AM, /obj/singularity) || istype(AM, /obj/item/projectile))
+		return
+	drop(AM)
+
+
+/turf/simulated/chasm/proc/drop(atom/movable/AM)
+	visible_message("[AM] falls into [src]!")
+	if(z+1 > world.maxz)
```
Comment (RemieRichards) [inline code/modules/mining/mine_turfs.dm]: see below, no.

**Comment by RemieRichards** [comment]
@Incoming5643 I think simple_animals have a var that states whether they fly, ```flying``` iirc, but ghosts and such don't, though technically I don't believe ghosts move, they don't trigger things on movement (which is why golem runes do a process() rather than a crossed()) iirc.

**Comment by Jalleo** [comment]
Actually this is rather intuitive you could create a child of this called voidspace (imagine a event where some of the tiles on station turn to this anyone walking into them just die.

**Comment by Dorsisdwarf** [comment]
@Jalleo using this on station would drop people into centcomm

**Comment by KingDairy** [comment]
dropping rocks into them to hear if there's a level below might be helpful

**Comment by Bawhoppen** [comment]
implementing rocks when

File: code/modules/mining/mine_turfs.dm (hunk 2 above)
Comment (KorPhaeron) [inline code/modules/mining/mine_turfs.dm]: Would making them vars, then having it set to straight down on New (if the mapper failed to set any) be acceptable?

Comment (RemieRichards) [inline code/modules/mining/mine_turfs.dm]: I guess, although I don't see why my idea isn't doable.

Comment (KorPhaeron) [inline code/modules/mining/mine_turfs.dm]: It is, just wondering what it should default to if someone forgets to assign the vars

Comment (KorPhaeron) [inline code/modules/mining/mine_turfs.dm]: Cant have it dropping people to null after all

Comment (RemieRichards) [inline code/modules/mining/mine_turfs.dm]: null is nicer than deletion, atleast they get to choke to death (or survive, I forget which happens)

**Comment by Supermichael777** [comment]
https://youtu.be/cdbYsoEasio?t=40
:scream:

**Comment by KingDairy** [comment]
ore is a rock
push the miner's haul into the bottomless pit
get pushed into the bottomless pit by the miner

**Comment by AlexanderUlanH** [comment]
I think anything that could move around in space should be able to manage to not fall down a pit.

**Comment by Zilenan91** [comment]
How about instead of it deleting mobs on the last floor you just make there be lava or something down there so it kills everything but ghosts.

**Comment by Bawhoppen** [comment]
basement zlevel when

**Comment by octareenroon91** [comment]
Would it be appropriate to build lattice/catwalk/plating over this turf like space?

Should the chasm target maybe be set on a z-level or area basis?

**Comment by KorPhaeron** [comment]
@RemieRichards 

I did the thing

**Comment by RemieRichards** [comment]
thing did.

**Comment by Strathcona** [comment]
I think z-level or area basis would be much better. It opens up so many interesting mapping possibilities and doesn't have clunky reliance of z-level numbering. Image creating an escher-like world where you can fall in infinite loops. Or having a great many chasms dropping you in the same area so that the chasm and the zone don't have to be the same size. If you really want a chasm to kill instead of teleporting, make sure to simply put something deadly wherever you plop the person out.

**Comment by duncathan** [comment]
@Strathcona the vars he added should make that possible.

**Pull request merged**